### PR TITLE
ZK-4377: Provide a Linelayout Component

### PR DESCRIPTION
**Pull request by plijyfes**

### CODE REVIEW

File: zktest/src/archive/test2/F90-ZK-4377.zul [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	<button label="setHflex()" onClick='item3.setHflex("")'></button>
+	<button label="setHeight(300px)" onClick='item3.setHeight("300px")'></button>
+	<button label="setHeight(60%)" onClick='item3.setHeight("60%")'></button>
+	<button label="setHeight(null)" onClick='item3.setHeight(null)'></button>
+	<button label="setWidth(300px)" onClick='item3.setWidth("300px")'></button>
+	<button label="setWidth(60%)" onClick='item3.setWidth("60%")'></button>
+	<button label="setWidth()" onClick='item3.setWidth("")'></button>
+	<button label="getIndex" onClick='Clients.log("index:" + item3.getIndex())'></button>
+	<linelayout id="target">
+		<lineitem pointIconSclass="z-icon-times">
```
Comment (scribetw) [inline zktest/src/archive/test2/F90-ZK-4377.zul]: test `<template>`

File: zul/src/archive/metainfo/xml/zul.xsd [hunk 2]
```diff
@@ -2190,6 +2194,32 @@
...
+		<xs:attributeGroup ref="rootAttrGroup" />
+	</xs:complexType>
+	<!-- lineitem -->
+	<xs:element name="lineitem" type="lineitemType" />
+	<xs:complexType name="lineitemType">
+		<xs:group ref="baseGroup" minOccurs="0" maxOccurs="unbounded" />
+		<xs:attributeGroup ref="xulElementAttrGroup" />
+		<xs:attribute name="pointVisible" type="booleanType" use="optional" />
+		<xs:attribute name="pointStyle" type="xs:string" use="optional" />
+		<xs:attribute name="pointSrc" type="xs:string" use="optional" />
```
Comment (scribetw) [inline zul/src/archive/metainfo/xml/zul.xsd]: pointImage, pointImageContent // LabelImageElement

File: zul/src/archive/metainfo/xml/zul.xsd [hunk 3]
```diff
@@ -2190,6 +2194,32 @@
...
 		<xs:attributeGroup ref="rootAttrGroup" />
 	</xs:complexType>
+	<!-- linelayout -->
+	<xs:element name="linelayout" type="linelayoutType" />
+	<xs:complexType name="linelayoutType">
+		<xs:group ref="baseGroup" minOccurs="0" maxOccurs="unbounded" />
+		<xs:attributeGroup ref="xulElementAttrGroup" />
+		<xs:attribute name="orient" type="xs:string" use="optional" />
+		<xs:attribute name="lineStyle" type="xs:string" use="optional" />
+		<xs:attribute name="firstScale" type="xs:string" use="optional" />
```
Comment (scribetw) [inline zul/src/archive/metainfo/xml/zul.xsd]: int

File: zktest/src/org/zkoss/zktest/test2/F90_ZK_4377_2Composer.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+			listIterator.next();
+			listIterator.set("changed!");
+		}
+	}
+	
+	@Listen("onClick=#setModelABC")
+	public void setModelABC() {
+		model = new ListModelList<>(Arrays.asList("A", "B", "C"));
+		lineLayout.setModel(model);
+	}
```
Comment (plijyfes) [inline zktest/src/org/zkoss/zktest/test2/F90_ZK_4377_2Composer.java]: test setModel(null)

File: zktest/src/archive/test2/F90-ZK-4377.zul [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	<button label="setWidth(2000px)" onClick='target.setWidth("2000px")'></button>
+	<button label="setHeight(100%)" onClick='target.setHeight("100%")'></button>
+	<button label="setHeight(2000px)" onClick='target.setHeight("2000px")'></button>
+	<separator/>
+	<label>Lineitem3:</label>
+	<button label="setPointVisible(false)" onClick="item3.setPointVisible(false)"></button>
+	<button label="setPointVisible(true)" onClick="item3.setPointVisible(true)"></button>
+	<button label="setPointImageSrc" onClick='item3.setPointImageSrc("/test2/img/sun.jpg")'></button>
+	<button label="clearPointImageSrc" onClick='item3.setPointImageSrc("")'></button>
+	<button label="setPointImageContent" onClick='item3.setPointImageContent(new org.zkoss.image.AImage("t", desktop.webApp.getResourceAsStream("/test2/img/inet.png")))'></button>
```
Comment (plijyfes) [inline zktest/src/archive/test2/F90-ZK-4377.zul]: attribute name="onClick"

File: zktest/src/archive/test2/F90-ZK-4377-2.zul [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+<zk>
+	<label multiline="true">
+		Try these buttons for model and template test:
+	</label>
+	<div apply="org.zkoss.zktest.test2.F90_ZK_4377_2Composer" width="100%" height="100%">
+		<button label="add item" id="addItem" />
+		<button label="remove item" id="removeItem" />
+		<button label="change item" id="changeItem" />
+		<button label="setModelABC" id="setModelABC" />
+		<linelayout id="lineLayout" orient="horizontal">
```
Comment (plijyfes) [inline zktest/src/archive/test2/F90-ZK-4377-2.zul]: test bindrenderer MVVM

File: zktest/src/archive/test2/config.properties [hunk 7]
```diff
@@ -3189,6 +3189,9 @@ F86-ZK-4235.zul=A,E,datefmt,library-property
 ##manually##F90-ZK-4386.zul=A,E,Anchornav
 ##manually##F90-ZK-4379.zul=A,E,Loadingbar
 ##manually##F90-ZK-4395.zul=A,E,Pdfviewer
+##manually##F90-ZK-4377.zul=A,E,Linelayout
+##zats##F90-ZK-4377-2.zul=A,E,Linelayout
+##zats##F90-ZK-4377-3.zul=A,E,Linelayout
```
Comment (plijyfes) [inline zktest/src/archive/test2/config.properties]: ,MVVM

File: zktest/src/org/zkoss/zktest/test2/F90_ZK_4377_3VM.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	public ListModel<String> getData() {
+		return data;
+	}
+	
+	public void setData(ListModelList<String> dataList) {
+		this.data = dataList;
+	}
+	
+	@Command("add")
+	@NotifyChange("data")
```
Comment (plijyfes) [inline zktest/src/org/zkoss/zktest/test2/F90_ZK_4377_3VM.java]: remove @NotifyChange("data")

**Pull request merged**